prov/efa/src/efa_av.c

```diff
@@ -1075,3 +1075,102 @@ int efa_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
 	free(av);
 	return ret;
 }
+
+#if EFA_UNIT_TEST
+
+#include "efa_unit_tests.h"
+
+/* Common Setup/Teardown */
+
+struct efa_resource {
+	struct fid_fabric *fabric;
+	struct fid_domain *domain;
+	struct fid_av *av;
+};
+
+struct efa_resource *create_efa_resource() {
+	struct efa_resource *resource = (struct efa_resource*) malloc(sizeof(struct efa_resource));
+	struct fi_info *info;
+	struct fi_av_attr av_attr = {0};
+	int err;
+
+	resource->fabric = NULL;
+	resource->domain = NULL;
+	resource->av = NULL;
+
+	err = fi_getinfo(FI_VERSION(1, 14), NULL, NULL, 0ULL, NULL, &info);
+	if (err) {
+		free(resource);
+		return NULL;
+	}
+
+	err = fi_fabric(info->fabric_attr, &resource->fabric, NULL);
+	if (err) {
+		free(resource);
+		return NULL;
+	}
+
+	err = fi_domain(resource->fabric, info, &resource->domain, NULL);
+	if (err) {
+		fi_close(&resource->fabric->fid);
+		free(resource);
+		return NULL;
+	}
+
+	err = fi_av_open(resource->domain, &av_attr, &resource->av, NULL);
+	if (err) {
+		fi_close(&resource->domain->fid);
+		fi_close(&resource->fabric->fid);
+		free(resource);
+		return NULL;
+	}
+
+	return resource;
+}
+
+void delete_efa_resource(struct efa_resource *resource) {
+	fi_close(&resource->av->fid);
+	fi_close(&resource->domain->fid);
+	fi_close(&resource->fabric->fid);
+	free(resource);
+}
+
+/*
+ * Only works on nodes with EFA devices
+ * This test calls efa_ah_alloc twice with the same GID,
+ * and verifies that ibv_create_ah only gets called once.
+ */
+void test_duplicate_efa_ah_creation() {
+	/* Setup Local Variables */
+	struct efa_resource *resource = create_efa_resource();
```

Comment:
If create_efa_resource returns NULL because fi_get_info returns an error, should we handle that and not move forward in that case?

Reply:
I had an assert lower down `assert_non_null(resource);`. It got a little messy when I was shuffling definitions to the top of the function.  I took the separated declaration and assignment of resource to make this more clear.